config/eventvalidation.go

```diff
@@ -0,0 +1,138 @@
+	Start:         {},
+	FirstQuartile: {},
+	MidPoint:      {},
+	ThirdQuartile: {},
+	Complete:      {},
+}
+
+// validate verifies the events object  and returns error if at least one is invalid.
+func (e Events) validate(errs []error) []error {
+	if e.Enabled { // validate only if events are enabled
+		if !isValidURL(e.DefaultURL) {
+			return append(errs, errors.New("Invalid events.default_url"))
+		}
+		err := validateVASTEvents(e.VASTEvents)
+		if err != nil {
+			return append(errs, err)
+		}
+	}
+	return errs
+}
+
+// validateVASTEvents verifies the all VASTEvent objects and returns error if at least one is invalid.
+func validateVASTEvents(events []VASTEvent) error {
+	if events != nil {
+		for i, event := range events {
+			if err := validateVASTEvent(event, i); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// validateVASTEvent validates event object and  returns error if at least one is invalid
+func validateVASTEvent(event VASTEvent, index int) error {
+	if !isValidCreateElement(event.CreateElement) {
+		return fmt.Errorf("Invalid events.vast_events[%d].create_element", index)
+	}
+
+	// VASTEvent.ExcludeDefaultURL assumed to be false by default
+	if !isValidType(event) {
+		if isTrackingEvent(event) {
+			var ele []string
+			for k := range vastEventElementMap {
+				ele = append(ele, string(k))
+			}
+			return fmt.Errorf("Missing or Invalid events.vast_events[%d].type. Valid values are %v", index, strings.Join(ele, ", "))
+		}
+		return fmt.Errorf("events.vast_events[%d].type is not applicable for create element '%s'", index, event.CreateElement)
+	}
+	for i, url := range event.URLs {
+		if !isValidURL(url) {
+			return fmt.Errorf("Invalid events.vast_events[%d].urls[%d]", index, i)
+		}
+	}
+	// ensure at least one valid url exists when default URL to be excluded
+	if event.ExcludeDefaultURL && len(event.URLs) == 0 {
+		return fmt.Errorf("Please provide at least one valid URL in events.vast_events[%d].urls or set events.vast_events[%d].exclude_default_url=false", index, index)
+	}
+
+	return nil // no errors
+}
+
+// isValidCreateElement checks create_element has valid value
+// if value is value returns true, otherwise false
+func isValidCreateElement(element VASTEventElement) bool {
+	// validate create element
+	if _, ok := vastEventElementMap[element]; ok {
+		return true
+	}
+	return false
+}
+
+// isValidtype checks if valid type is provided in case event is of type tracking.
+// in case of other events this value must be empty
+func isValidType(event VASTEvent) bool {
+	if isTrackingEvent(event) {
+		if _, ok := trackingEventTypeMap[event.Type]; ok {
+			// valid event type for create element tracking
+			return true
+		}
+		return false
+	}
+	return len(event.Type) == 0 // event.type must be empty in case create element is not tracking
+}
+
+// isValidURL validates the event URL
+func isValidURL(eventURL string) bool {
+	return validator.IsURL(eventURL) && validator.IsRequestURL(eventURL)
```

Comment:
Using both IsURL and IsRequestURL because IsURL allows relative paths
hereas IsRequestURL requires absolute path but fails to check other valid URL
format constraints.

For example: IsURL will allow "abcd.com" but IsRequestURL won't
IsRequestURL will allow "http://http://abcd.com" but IsURL won't

This is referred from
1.  Same is being used here -   https://github.com/prebid/prebid-server/blob/2590c682fa8cde86b18496ee35cd546c011e3f92/config/adapter.go#L77